automata/pda/npda.py

```diff
@@ -111,6 +125,44 @@ def _get_next_configurations(
             new_configs.add(new_config)
         return new_configs
 
+    def _get_input_path(
+        self, input_str: str
+    ) -> Tuple[List[Tuple[PDAConfiguration, PDAConfiguration]], bool]:
+        """
+        Calculate the path taken by input.
+
+        Args:
+            input_str (str): The input string to run on the NPDA.
+
+        Returns:
+            Tuple[List[Tuple[PDAConfiguration, PDAConfiguration]], bool]: A list
+            of all transitions taken in each step and a boolean indicating
+            whether the NPDA accepted the input.
+
+        """
+
+        steps = list(self.read_input_stepwise(input_str))
+
+        path: List[PDAConfiguration] = [steps.pop().pop()]
+
+        accepted = path[0] in self.final_states
+
+        for i in reversed(steps):
+            if len(i) == 1:
+                path.append(i.pop())
+                continue
+
+            for curr_step in i:
+                for next_step in self._get_next_configurations(curr_step):
+                    if next_step == path[-1]:
+                        path.append(curr_step)
+                        break
```

Comment:
A little bit of a nitpick, but I think now there are tests for this and it has already been simplified somewhat, it would be good to see if we could simplify these cases a little bit before merging.

Reply:
I believe it is simple enough. Stared at the code for 2 mins and could not think of anything. Do you have anything in mind to simplify this.

Reply:
So the `_get_next_configurations(curr_step)` function returns a set, and you're just checking whether `path[-1]` is contained in there. So I think you can just do `if path[-1] in self._get_next_configurations(curr_step)` instead of the innermost loop, and then do `path.append(curr_step)` inside the new if block.

Reply:
@eliotwrobson Sorry, for the late response. I have implemented your suggestions. Please have a look. The code is way easier to read with this modification you specified. ☺️

Reply:
@eliotwrobson Please take a look at @Vipul-Cariappa's changes, and if you are satisfied, feel free to resolve this conversation thread.

Reply:
New code looks great! Will resolve and thanks for your patience @Vipul-Cariappa